streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimeWindowStoreMaterializer.java

```diff
@@ -0,0 +1,133 @@
+    ) {
+        super(materialized);
+        this.windows = windows;
+        this.emitStrategy = emitStrategy;
+    }
+
+    @Override
+    public StateStore build() {
+        WindowBytesStoreSupplier supplier = (WindowBytesStoreSupplier) materialized.storeSupplier();
+        if (supplier == null) {
+            final long retentionPeriod = retentionPeriod();
+
+            if ((windows.size() + windows.gracePeriodMs()) > retentionPeriod) {
+                throw new IllegalArgumentException("The retention period of the window store "
+                        + materialized.storeName() + " must be no smaller than its window size plus the grace period."
+                        + " Got size=[" + windows.size() + "],"
+                        + " grace=[" + windows.gracePeriodMs() + "],"
+                        + " retention=[" + retentionPeriod + "]");
+            }
+
+            switch (defaultStoreType) {
+                case IN_MEMORY:
+                    supplier = Stores.inMemoryWindowStore(
+                            materialized.storeName(),
+                            Duration.ofMillis(retentionPeriod),
+                            Duration.ofMillis(windows.size()),
+                            false
+                    );
+                    break;
+                case ROCKS_DB:
+                    supplier = emitStrategy.type() == EmitStrategy.StrategyType.ON_WINDOW_CLOSE ?
+                            RocksDbIndexedTimeOrderedWindowBytesStoreSupplier.create(
+                                    materialized.storeName(),
+                                    Duration.ofMillis(retentionPeriod),
+                                    Duration.ofMillis(windows.size()),
+                                    false,
+                                    false
+                            ) :
+                            Stores.persistentTimestampedWindowStore(
+                                    materialized.storeName(),
+                                    Duration.ofMillis(retentionPeriod),
+                                    Duration.ofMillis(windows.size()),
+                                    false
+                            );
+                    break;
+                default:
+                    throw new IllegalStateException("Unknown store type: " + materialized.storeType());
+            }
+        }
+
+        final StoreBuilder<TimestampedWindowStore<K, V>> builder = Stores.timestampedWindowStoreBuilder(
+                supplier,
+                materialized.keySerde(),
+                materialized.valueSerde()
+        );
+
+        if (materialized.loggingEnabled()) {
+            builder.withLoggingEnabled(materialized.logConfig());
+        } else {
+            builder.withLoggingDisabled();
+        }
+
+        // TODO(agavra): remove before merging, should we do what we do with other stores
+        // and disable caching in the case StrategyType.ON_WINDOW_CLOSE is used?
```

Comment:
@ableegoldman @guozhangwang do any of you know whether we should do this? is not having it here but having it in the other two stores a bug?

Reply:
Would be good to have @guozhangwang  confirm, or maybe @mjsax or @lihaosky can chime in here, but my understanding is that we have to disable caching when using `ON_WINDOW_CLOSE` because the alternative store implementation has a different ordering scheme than the CachingWindow/SessionStore, breaking range queries as we can't properly merge results between the cache and underlying store.

Of course in the DSL, sliding and session window aggregations use range queries whereas time window aggregations do not. And since this seems to be a DSL-specific feature, I guess we just don't care about the incompatible cache ordering for time windows since we'll only be doing point lookups.

I also noticed that we added a new caching layer specifically to work with the RocksDbIndexedTimeOrderedWindowBytesStore, called [TimeOrderedCachingWindowStore](https://github.com/apache/kafka/pull/12030/files#diff-65aaedea554a4017e0e1a673883e3b73f609f117f71e86541c9d45092bf0dbd2R59) Not sure how that comes into play here, but I think it makes it clear that we do intend to allow caching with time windows (only)

Code-wise, the difference between sliding windows and time windows seems to be that the former is "indexed", as you can see by following the invocations of `RocksDbIndexedTimeOrderedWindowBytesStoreSupplier#create` Presumably this means the TimeOrderedCachingWindowStore doesn't work for sliding window stores, only time window stores. 

I suppose it would be nice to include a condensed version of this in the comments about this in the session and sliding window materializers. Right now they only state that caching _is_ disabled (which we can plainly see from the code), but not _why_ it is disabled, which is the important part that can't be understood from the code.

Reply:
sounds good, thanks for the detailed explanation @ableegoldman! For now I just removed the TODO since I'm not sure I fully understand the nuance here enough to add the comment on my own. It'd be good if either @mjsax or @lihaosky had a quick follow-up PR to add the comments with the inline explanation.

Reply:
Needed to dig some code to refresh my memory (which is still not totally up-to-date). I found https://github.com/apache/kafka/pull/12204#discussion_r907968103

Guess it's ok to not have an explicit comment about it?

Reply:
Yeah it's not essential to have a comment, certainly shouldn't block this PR, but there is clearly a lot of subtle and case-by-case handling here that seems all too easy to mess up by accident in the future.

 Like that comment you just linked Matthias -- hadn't seen that before, is it saying we handle the cache differently depending on if it's persistent or not? I know the in-memory stores have a different layout than the rocksdb ones, but that's what the `SegmentedCacheFunction` is for in the MergedSortedCacheXXXIterators -- to account for the segmentation of rocksdb stores

I'm just wondering if we're doing something weird based on what's returned by `StateStore#persistent` -- we have a lot of what I would consider "bugs" in Streams due to assuming that `persistent` means RocksDBStore and  `!persistent` means InMemoryStore. We don't always account for custom stores.